R/utilities-simulation.R

```diff
@@ -749,3 +749,230 @@ getSimulationTree <- function(simulationOrFilePath, quantityType = "Quantity") {
+                           simulationRunOptions = NULL) {
+  ospsuite.utils::validateIsOfType(simulations, type = "Simulation")
+  ospsuite.utils::validateIsString(quantitiesPaths, nullAllowed = TRUE)
+  ospsuite.utils::validateIsNumeric(steadyStateTime, nullAllowed = FALSE)
+  simulations <- ospsuite.utils::toList(simulations)
+
+  if (any(steadyStateTime <= 0)) {
+    stop(messages$steadyStateTimeNotPositive(steadyStateTime))
+  }
+
+  # If `steadyStateTime` is a vector of values, it must be of the same size as
+  # the list of simulations.
+  # Otherwise, repeat the value for the number of simulations
+  if (length(steadyStateTime) > 1) {
+    ospsuite.utils::validateIsSameLength(simulations, steadyStateTime)
+  } else {
+    steadyStateTime <- rep(steadyStateTime, length(simulations))
+  }
+
+  # First prepare all simulations by setting their outputs and time intervals
+  # If no quantities have been specified, the quantities paths may be different
+  # for each simulation and must be stored separately
+  simulationState <- .storeSimulationState(simulations)
+  quantitiesPathsMap <- vector(mode = "list", length = length(simulations))
+  for (idx in seq_along(simulations)) {
+    simulation <- simulations[[idx]]
+    simId <- simulation$id
+    # Set simulation time to the steady-state value.
+    ospsuite::clearOutputIntervals(simulation = simulation)
+    simulation$outputSchema$addTimePoints(timePoints = steadyStateTime[[idx]])
+    # If no quantities are explicitly specified, simulate all outputs.
+    if (is.null(quantitiesPaths)) {
+      quantitiesPathsMap[[idx]] <- ospsuite::getAllStateVariablesPaths(simulation)
+    } else {
+      quantitiesPathsMap[[idx]] <- quantitiesPaths
+    }
+    names(quantitiesPathsMap)[[idx]] <- simId
+    ospsuite::clearOutputs(simulation)
+    ospsuite::addOutputs(quantitiesOrPaths = quantitiesPathsMap[[idx]], simulation = simulation)
+  }
+
+  # Run simulations concurrently
+  simulationResults <- ospsuite::runSimulations(
+    simulations = simulations,
+    simulationRunOptions = simulationRunOptions
+  )
+
+  # Iterate through simulations and get their outputs
+  outputMap <- vector(mode = "list", length = length(simulations))
+  for (idx in seq_along(simulations)) {
+    simulation <- simulations[[idx]]
+    simId <- simulation$id
+    simResults <- simulationResults[[simId]]
+
+    allOutputs <- ospsuite::getOutputValues(
+      simResults,
+      quantitiesOrPaths = quantitiesPathsMap[[simId]],
+      stopIfNotFound = stopIfNotFound,
+      addMetaData = FALSE
+    )
+
+    # Get the end values of all outputs
+    endValues <- lapply(quantitiesPathsMap[[simId]], function(path) {
+      # Check if the quantity is defined by an explicit formula
+      isFormulaExplicit <- ospsuite::isExplicitFormulaByPath(
+        path = enc2utf8(path),
+        simulation = simulation,
+        stopIfNotFound = stopIfNotFound
+      )
+
+      if (ignoreIfFormula && isFormulaExplicit) {
```

Comment:
I don't get this: how is a state variable supposed to be defined by explicit formula???

Reply:
What we are dealing here with is the **start value** of the parameter, which can be an explicit formula.

Reply:
ok... but I still don't understand the use case. So if a parameter is a state variable with initial value defined by formula - why would I need to ignore this parameter in the steady state?

Reply:
I sometimes define the start value of the parameter as a formula that calculates the steady-state of this parameter based on other intputs. In this case, I would not like to overwrite the start value.

Reply:
While I still don't understand the use case (maybe you can explain it to me during one of the meetings) - I wonder if we should set the default value of `ignoreIfFormula `to **FALSE**?
It seems pretty dangerous for me when the steady state of an ODE species is ignored because its initial value is not const. but defined by the formula.

Reply:
I have been thinking about it for some time now, and I am still not convinced, maybe we should discuss it in a meeting.

The way I see it, if a start value of an ODE parameter (and the same goes for molecules, e.g. start values of proteins) is defined by a formula, then this formula represents the steady-state. Otherwise, there is something off with the equation, don't you think?